### ISSUE TITLE
How much luck is involved when it comes to training a YOLOv5 model?

### ISSUE THREAD

**Issue by haimat** [ultralytics/yolov5#9262]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have trained a YOLOv5-X model on a custom dataset in 500 epochs to a mAP value of 0.35. Then I started another training, using the same hyperparams, with the `--weights` parameter using the `best.pt` file from the first training, again with 500 epochs. I repeated that second training 5 times, every time exactly the same way - same dataset, same initial weights, same machine, same hyperparams.

In four of these 2nd training sessions I ended up again with a final mAP of around 0.35 again - either because I reached last training epoch of 500, or because the `patience` value of YOLO stopped training. However, in one of these five training sessions I ended up with a mAP value of 0.55 - way more than all the other trainings.

Now since I changed absolutely nothing, simply re-tried the same experiment for 5 times, I am wondering how we can know that the optimal model performance has been reached? I mean, if I would have tried the 2nd training session only once, then I would not have seen that much better training up to mAP = 0.55, but would still stick to a mAP value of 0.35.

Is there a better way then just trying the same training multiuple times?

**Comment by glenn-jocher** [comment]
@haimat 👋 Hello! Thanks for asking about **training reproducibility**. YOLOv5 🚀 uses a single training seed which is set here using the `init_seeds()` function. CPU and Single-GPU trainings should be fully reproducible with `torch>=1.12.0`. Multi-GPU DDP trainings are still not reproducible unfortunately. This is an open issue for us and we could use any help in tracking down this problem.
https://github.com/ultralytics/yolov5/blob/7215a0fb41a90d8a0bf259fa708dff608a1f0262/train.py#L104

This function sets `python`, `numpy` and `torch` seeds and updates `cudnn` settings:
https://github.com/ultralytics/yolov5/blob/7215a0fb41a90d8a0bf259fa708dff608a1f0262/utils/general.py#L198-L214

To set a new training seed for example:
```bash
python train.py --seed 3  # default seed=0
```

Note that even when using the same seed trainings may produce different results, especially when using CUDA backends. See https://pytorch.org/docs/stable/notes/randomness.html for details on factors affecting PyTorch training reproducibility and sources of randomness.

<img width="949" alt="Screen Shot 2022-01-08 at 11 20 59 AM" src="https://user-images.githubusercontent.com/26833433/148660366-9155dfd9-351c-4174-818f-2b1b907b3e4e.png">

Good luck 🍀 and let us know if you have any other questions!
